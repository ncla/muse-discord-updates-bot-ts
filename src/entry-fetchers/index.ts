import {Update} from "@/src/updates";

export interface EntryFetcher {
    fetch(): Promise<Update[]>;
}