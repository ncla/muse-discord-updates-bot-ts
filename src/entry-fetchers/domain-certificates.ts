import {EntryFetcher} from "@/src/entry-fetchers/index";
import {z} from "zod";
import {createBlankUpdate, DomainCertificateUpdate, UpdateType} from "@/src/updates";

const CertificateSchema = z.object({
    issuer_ca_id: z.number(),
    issuer_name: z.string(),
    common_name: z.string(),
    name_value: z.string(),
    id: z.number(),
    entry_timestamp: z.string(),
    not_before: z.string(),
    not_after: z.string(),
    serial_number: z.string(),
    result_count: z.number(),
})

type Certificate = z.infer<typeof CertificateSchema>

export class DomainCertificates implements EntryFetcher
{
    private domain: string | undefined;

    constructor(domain: string | undefined) {
        this.domain = domain
    }

    async fetch()
    {
        if (this.domain === undefined) {
            throw new Error('Domain is not set')
        }

        const url = new URL(`https://crt.sh/json`);
        url.searchParams.append('q', this.domain)

        const response = await fetch(url.toString());

        if (!response.ok) {
            throw new Error(`Response status: ${response.status}`);
        }

        const json = await response.json();

        const CertificateArraySchema = z.array(CertificateSchema)

        const parsedJson = CertificateArraySchema.parse(json)

        const uniqueCommonNames = new Set(
            parsedJson
                .map((cert: Certificate) => cert.name_value.split('\n'))
                .flat()
        )

        return Array.from(uniqueCommonNames).map(commonName => this.mapCertificateToUpdateEntry(commonName))
    }

    private mapCertificateToUpdateEntry(domain: string): DomainCertificateUpdate
    {
        return {
            ...createBlankUpdate(),
            type: UpdateType.DOMAIN_CERTIFICATE,
            id: domain,
            uniqueId: domain,
        }
    }
}