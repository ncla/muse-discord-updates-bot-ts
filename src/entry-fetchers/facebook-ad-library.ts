import { EntryFetcher } from "@/src/entry-fetchers";
import * as puppeteer from 'puppeteer';
import { FacebookAdUpdate, UpdateType } from "@/src/updates";
import * as fs from 'fs/promises';
import * as path from 'path';
import * as os from 'os';
import * as Sentry from "@sentry/node";
import { scrollUntilNoMoreContentLoads } from "@/src/common";

export class FacebookAdLibrary implements EntryFetcher {
    private readonly baseUrl: string = 'https://www.facebook.com/ads/library/?active_status=all&ad_type=all&country=ALL&is_targeted_country=false&media_type=all&search_type=page';

    constructor(private readonly pageId: string | undefined) {}

    async fetch(): Promise<FacebookAdUpdate[]> {
        if (!this.pageId) {
            console.log('No Facebook page ID set for Ad Library fetcher');
            return [];
        }
        
        const url = `${this.baseUrl}&view_all_page_id=${this.pageId}`;
        const browser = await puppeteer.launch({
            // headless: false
        });

        try {
            const page = await browser.newPage();

            await page.setViewport({ width: 1280, height: 800 });

            console.log(`Navigating to Facebook Ad Library: ${url}`);
            
            await page.goto(url, {
                waitUntil: 'networkidle0',
                timeout: 60000
            });

            console.log('Scrolling down to load more ads...');
            
            await scrollUntilNoMoreContentLoads(
                page,
                null,
                '[data-visualcompletion="loading-state"]',
                'end'
            );
            
            console.log('Finding ad containers...');
            const adContainers = await this.findAdContainers(page);
            console.log(`Found ${adContainers.length} ad containers`);

            if (adContainers.length === 0) {
                console.log('No ads found, returning empty array');
                return [];
            }
            
            console.log(`Processing all ${adContainers.length} containers`);
            
            const foundIds = new Set<string>();
            const updates: FacebookAdUpdate[] = [];
            
            for (let i = 0; i < adContainers.length; i++) {
                try {
                    const container = adContainers[i];

                    await this.hideFixedElements(page);

                    const screenshotPath = await this.takeContainerScreenshot(container, i);
                    
                    const adId = await this.extractLibraryIdFromContainer(container, i);
                    
                    if (!adId) {
                        console.log(`No Library ID found for container ${i}, skipping`);
                        continue;
                    }
                    
                    if (foundIds.has(adId)) {
                        console.log(`Duplicate Library ID ${adId} found in container ${i}, skipping`);
                        continue;
                    }
                    
                    foundIds.add(adId);
                    
                    updates.push({
                        type: UpdateType.FACEBOOK_AD,
                        uniqueId: adId,
                        id: adId,
                        screenshot: screenshotPath,
                    });
                } catch (error) {
                    console.error(`Error processing ad container ${i}:`, error);
                }
            }

            console.log(`Successfully extracted ${updates.length} unique ads`);
            return updates;
        } catch (e) {
            console.error('Error fetching Facebook Ad Library:', e);
            Sentry.captureException(e);
            return [];
        } finally {
            await browser.close();
            const process = browser.process();
            if (process && !process.killed) {
                process.kill('SIGTERM');
            }
        }
    }
    
    private async extractLibraryIdFromContainer(container: puppeteer.ElementHandle<Element>, index: number): Promise<string | null> {
        try {
            const libraryId = await container.evaluate(el => {
                const text = el.textContent || '';
                const match = text.match(/Library ID:\s*(\d+)/i);
                return match && match[1] ? match[1] : null;
            });
            
            if (libraryId) {
                console.log(`Found Library ID: ${libraryId} in container ${index}`);
                return libraryId;
            }
            
            return null;
        } catch (error) {
            console.error(`Error extracting Library ID from container ${index}:`, error);
            return null;
        }
    }

    private async findAdContainers(page: puppeteer.Page): Promise<puppeteer.ElementHandle<Element>[]> {
        const matchCount = await page.evaluate(() => {
            const divs = document.querySelectorAll('div');
            let matchCount = 0;
            
            for (let i = 0; i < divs.length; i++) {
                const div = divs[i];
                const style = window.getComputedStyle(div);
                const rect = div.getBoundingClientRect();
                
                if (style.display === 'none' || 
                    style.visibility === 'hidden' || 
                    rect.width < 300 || 
                    rect.height < 200) {
                    continue;
                }
                
                // Use positive matching for specific CSS properties
                let stylesMatched = 0;
                const requiredMatches = 11;
                
                // Border radius properties
                if (style.borderTopLeftRadius === '6px') stylesMatched++;
                if (style.borderTopRightRadius === '6px') stylesMatched++;
                if (style.borderBottomLeftRadius === '6px') stylesMatched++;
                if (style.borderBottomRightRadius === '6px') stylesMatched++;
                
                // Border properties
                if (style.borderTopWidth === '1px' && 
                    style.borderTopStyle === 'solid' &&
                    style.borderTopColor.includes('rgba(0, 0, 0, 0.15)')) stylesMatched++;
                
                if (style.borderLeftWidth === '1px' && 
                    style.borderLeftStyle === 'solid' &&
                    style.borderLeftColor.includes('rgba(0, 0, 0, 0.15)')) stylesMatched++;
                
                if (style.borderRightWidth === '1px' && 
                    style.borderRightStyle === 'solid' &&
                    style.borderRightColor.includes('rgba(0, 0, 0, 0.15)')) stylesMatched++;
                
                if (style.borderBottomWidth === '1px' && 
                    style.borderBottomStyle === 'solid' &&
                    style.borderBottomColor.includes('rgba(0, 0, 0, 0.15)')) stylesMatched++;
                
                // Box shadow
                if (style.boxShadow.includes('0px 0px 5px') && 
                    style.boxShadow.includes('rgba(0, 0, 0, 0.1)')) stylesMatched++;
                
                // Background color
                if (style.backgroundColor === 'rgb(255, 255, 255)') stylesMatched++;
                
                // Box sizing
                if (style.boxSizing === 'border-box') stylesMatched++;
                
                // Check if container has Library ID text
                if (stylesMatched === requiredMatches) {
                    const text = div.textContent || '';
                    const hasLibraryId = /Library ID:\s*\d+/i.test(text);
                    
                    if (hasLibraryId) {
                        div.setAttribute('data-ad-container', 'true');
                        console.log(`Container at (${rect.left}, ${rect.top}) with size ${rect.width}x${rect.height}, matched ${stylesMatched}/${requiredMatches} required styles and contains Library ID`);
                        matchCount++;
                    }
                }
            }
            
            return matchCount;
        });
        
        console.log(`Found ${matchCount} ad containers with Library ID`);
        
        const containers = await page.$$('[data-ad-container="true"]');
        
        console.log(`Successfully found ${containers.length} container elements`);
        return containers;
    }

    private async takeContainerScreenshot(
        container: puppeteer.ElementHandle<Element>,
        index: number
    ): Promise<string | undefined> {
        try {
            const tempDir = path.join(os.tmpdir(), 'muse-discord-bot');
            await fs.mkdir(tempDir, { recursive: true });
            
            const screenshotPath = path.join(tempDir, `facebook-ad-${Date.now()}-${index}.png`);
            await container.screenshot({ path: screenshotPath });
            console.log(`Captured screenshot of container ${index} to ${screenshotPath}`);
            
            return screenshotPath;
        } catch (error) {
            console.log(`Failed to take screenshot of container ${index}: ${(error as Error).message}`);
            return undefined;
        }
    }

    private async hideFixedElements(page: puppeteer.Page): Promise<void> {
        await page.evaluate(() => {
            // Select all fixed position elements
            const fixedElements = Array.from(document.querySelectorAll('*')).filter(el => {
                const style = window.getComputedStyle(el);
                return style.position === 'fixed' || style.position === 'sticky';
            });
            
            // Hide them
            fixedElements.forEach(el => {
                (el as HTMLElement).style.visibility = 'hidden';
            });
        });
    }
}
