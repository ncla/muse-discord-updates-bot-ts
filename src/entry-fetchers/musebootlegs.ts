import {EntryFetcher} from "@/src/entry-fetchers/index";
import config, {IConfig} from "@/src/config";
import {JSDOM} from 'jsdom'
import {createBlankUpdate, Update, UpdateType} from "@/src/updates";

export class Musebootlegs implements EntryFetcher
{
    private config: IConfig;

    constructor(config: IConfig) {
        this.config = config;
    }

    async fetch()
    {
        if (
            this.config.services.musebootlegs.username === undefined ||
            this.config.services.musebootlegs.password === undefined ||
            this.config.services.musebootlegs.user_agent === undefined
        ) {
            throw new Error('MuseBootlegs username, password or user agent is not set')
        }

        const loginResponse = await this.sendLoginRequest(
            this.config.services.musebootlegs.username,
            this.config.services.musebootlegs.password,
            this.config.services.musebootlegs.user_agent
        )

        const cookies = loginResponse.headers.getSetCookie()

        if (
            cookies[0] === undefined ||
            !cookies[0].startsWith('tsue_member=')
        ) {
            throw new Error('Failed to get login cookies')
        }

        const cookieHeader = cookies.map(cookie => cookie.split(';')[0]).join('; ')

        const torrentListResponse = await this.sendTorrentListRequest(
            cookieHeader,
            this.config.services.musebootlegs.user_agent
        )

        if (!torrentListResponse.ok) {
            throw new Error('Failed to get torrent list response')
        }

        const torrentListPageHtml = await torrentListResponse.text()

        return this.parseTorrentListResponse(torrentListPageHtml)
    }

    async sendLoginRequest(
        username: string | undefined = undefined,
        password: string | undefined = undefined,
        useragent: string | undefined = undefined
    )
    {
        if (
            typeof username !== 'string' ||
            typeof password !== 'string' ||
            typeof useragent !== 'string'
        ) {
            throw new Error('Missing required parameters for login request')
        }

        const url = 'https://www.musebootlegs.com/ajax/login.php'
        const formData = new FormData()
        formData.append('action', 'login')
        formData.append('loginbox_remember', 'true')
        formData.append('loginbox_membername', username)
        formData.append('loginbox_password', password)

        return fetch(url, {
            method: 'POST',
            body: formData,
            headers: {
                // Special user agent must be passed to bypass Cloudflare's protection
                'User-Agent': useragent
            }
        })
    }

    async sendTorrentListRequest(
        cookie: string,
        useragent: string | undefined = undefined
    )
    {
        if (typeof useragent !== 'string') {
            throw new Error('MuseBootlegs user agent is not set')
        }

        const url = 'https://www.musebootlegs.com/?p=torrents&pid=10'
        const formData = new FormData()
        formData.append('sortOptions[sortBy]', 'added')
        formData.append('sortOptions[sortOrder]', 'desc')

        return fetch(url, {
            method: 'POST',
            body: formData,
            headers: {
                'Cookie': cookie,
                // Special user agent must be passed to bypass Cloudflare's protection
                'User-Agent': useragent
            }
        })
    }

    async parseTorrentListResponse(html: string)
    {
        const dom = new JSDOM(html)

        const document = dom.window.document;

        let errorBox = document.querySelector('#show_error');

        if (errorBox) {
            throw new Error('Error box found')
        }

        let torrentBoxes = document.querySelectorAll('#content > .torrent-box[id^="torrent_"]');

        const entries: Update[] = []

        torrentBoxes.forEach((torrentBox) => {
            const entryTextElement = torrentBox.querySelector('.newIndicator')

            if (entryTextElement === null) {
                return
            }

            if (entryTextElement.textContent === null) {
                return
            }

            const entryLink = torrentBox.querySelector('.newIndicator a') as HTMLAnchorElement
            const imageElement = torrentBox.querySelector('.previewImage a') as HTMLAnchorElement
            const authorElement = torrentBox.querySelector('.torrentOwner > span > span')
            const torrentOwnerElement = torrentBox.querySelector('.torrentOwner')
            const descriptionElement = torrentBox.querySelector('.torrentDescription')

            let uploadedDate: Date | null = null

            if (torrentOwnerElement !== null) {
                try {
                    torrentOwnerElement.childNodes.forEach((child) => {
                        // Hacky way to get TEXT_NODE constant https://github.com/jsdom/jsdom/issues/2993
                        if (child.nodeType === new JSDOM('').window.Node.TEXT_NODE) {
                            const uploadedText = child.textContent ?? ''

                            const timestamp = uploadedText.replace('Uploaded ', '').replace(' by', '').trim()
                            const parts = timestamp.split(/[- :]/)

                            uploadedDate = new Date(Date.UTC(
                                parseInt(parts[2]),
                                parseInt(parts[1]) - 1,
                                parseInt(parts[0]),
                                parseInt(parts[3]),
                                parseInt(parts[4])
                            ))
                        }
                    })
                } catch (e) {
                    console.error('Error parsing uploaded date:', e)
                }
            }

            const torrentId = torrentBox.id.replace(/^\D+/g, '')

            entries.push({
                ...createBlankUpdate(),
                type: UpdateType.MUSEBOOTLEGS_TORRENT,
                id: torrentId,
                uniqueId: torrentId,
                title: entryTextElement.textContent.trim(),
                content: descriptionElement && descriptionElement.textContent ? descriptionElement.textContent.trim() : null,
                url: entryLink && entryLink.href ? entryLink.href : null,
                image_url: imageElement && imageElement.href ? imageElement.href : null,
                author: {
                    id: null,
                    name: authorElement && authorElement.textContent ? authorElement.textContent : null,
                    image_url: null
                },
                created_at: uploadedDate
            });
        });

        return entries
    }
}