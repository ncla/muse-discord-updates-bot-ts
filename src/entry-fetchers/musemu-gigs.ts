import {EntryFetcher} from "@/src/entry-fetchers/index";
import * as puppeteer from 'puppeteer';
import {MuseMuGigsUpdate, UpdateType} from "@/src/updates";
import * as Sentry from "@sentry/node";

export class MusemuGigs implements EntryFetcher
{
    async fetch()
    {
        const browser = await puppeteer.launch({
            // headless: false,
        })

        const page = await browser.newPage()

        let reachedLastPage = false
        let pageNumber = 0
        let gigs: MuseMuGigsUpdate[] = []

        while (!reachedLastPage) {
            const url = `https://www.muse.mu/tour?page=${pageNumber}`

            await page.goto(url, {
                waitUntil: 'networkidle0'
            })

            const currentPageGigs = await this.parseTourPageElements(page)

            gigs = [...gigs, ...currentPageGigs]

            // Prevent too much DoS if the logic for determining last page is incorrect
            if (currentPageGigs.length === 0 || pageNumber === 10) {
                reachedLastPage = true
                break
            }

            pageNumber++
        }

        await browser.close()

        return gigs
    }

    async parseTourPageElements(page: puppeteer.Page): Promise<MuseMuGigsUpdate[]> {
        const selector = '.view-tour .view-content .item-list ul li'

        const listElements = await page.$$(selector)

        const results: MuseMuGigsUpdate[] = []

        for (const listEl of listElements) {
            try {
                const tourLinkElement = await listEl.$('.tourMoreInfoLink');
                const tourTitleElement = await listEl.$('.tourtitle')
                const tourCityElement = await listEl.$('.tourCity')
                const eventDateElement = await listEl.$('.event-date')

                if (!tourLinkElement || !tourTitleElement || !tourCityElement || !eventDateElement) {
                    continue
                }

                const url = await tourLinkElement.evaluate(el => el.getAttribute('href'))

                if (!url) {
                    continue
                }

                const tourTitleElText = await tourTitleElement.evaluate(el => el.textContent)

                if (!tourTitleElText) {
                    continue
                }

                const tourCityElText = await tourCityElement.evaluate(el => el.textContent)

                if (!tourCityElText) {
                    continue
                }

                const eventDateText = await eventDateElement.evaluate(el => el.textContent)
                const eventDateUTC = new Date(eventDateText + 'UTC')

                let title = `${tourTitleElText.trim()}, ${tourCityElText.trim()}`

                if (title.endsWith(',')) {
                    title = title.slice(0, -1)
                }

                results.push({
                    type: UpdateType.MUSEMU_GIG,
                    uniqueId: new URL(url, 'https://muse.mu').pathname.replace('/tour-date/', ''),
                    id: new URL(url, 'https://muse.mu').pathname.replace('/tour-date/', ''),
                    title: title,
                    url: new URL(url, 'https://muse.mu').href,
                    event_date: eventDateUTC,
                })
            } catch (error) {
                console.error('Error processing list element:', error);
                Sentry.captureException(error);
            }
        }

        return results
    }
}