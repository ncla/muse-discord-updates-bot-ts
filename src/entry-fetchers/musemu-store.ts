import {EntryFetcher} from "@/src/entry-fetchers";
import * as puppeteer from 'puppeteer';
import {MusemuStoreUpdate, UpdateType} from "@/src/updates";

export class MusemuStore implements EntryFetcher
{
    async fetch()
    {
        const browser = await puppeteer.launch({
            // headless: false
        });

        const page = await browser.newPage();
        
        await page.goto('https://store.muse.mu/eu/search/?q=', {
            waitUntil: 'networkidle0'
        });

        await this.scrollUntilNoMoreContentLoads(page, '.product-grid__bottom');

        const products = await this.parseSearchResults(page);

        await browser.close();

        console.log(`Found ${products.length} products`);

        return products;
    }

    private async scrollUntilNoMoreContentLoads(page: puppeteer.Page, targetSelector: string): Promise<void> {
        const startTime = Date.now();
        const MAX_DURATION = 20000;

        while (true) {
            // Check if we've exceeded the maximum time
            if (Date.now() - startTime >= MAX_DURATION) {
                console.log('Reached maximum scroll time limit of 20 seconds');
                break;
            }

            // Scroll so the target element just appears at the bottom of viewport
            await page.evaluate((selector) => {
                const element = document.querySelector(selector);
                if (element) {
                    const elementRect = element.getBoundingClientRect();
                    const elementTop = element.getBoundingClientRect().top + window.scrollY;

                    // Calculate scroll position where element would appear at bottom of viewport
                    // Subtract element's height to ensure we can see the whole element
                    const scrollPosition = elementTop - window.innerHeight + elementRect.height;
                    window.scrollTo({
                        top: scrollPosition,
                        behavior: 'smooth'
                    });
                } else {
                    // If element not found, scroll to bottom as fallback
                    window.scrollTo(0, document.body.scrollHeight);
                }
            }, targetSelector);

            try {
                await page.waitForSelector('.spinner', {
                    visible: true,
                    timeout: 1000
                });

                await page.waitForSelector('.spinner', {
                    hidden: true,
                    timeout: 10000
                });

            } catch (error) {
                if (error instanceof puppeteer.TimeoutError) {
                    console.log('No more content to load');
                    break;
                }

                throw error;
            }

            // Small pause to let any new content render
            await new Promise(r => setTimeout(r, 500));
        }
    }


    private async parseSearchResults(page: puppeteer.Page): Promise<MusemuStoreUpdate[]> {
        const productsFromPageContext = await page.evaluate(() => {
            return Array.from(document.querySelectorAll('.product-grid .product')).map(el => ({
                id: el.getAttribute('data-pid'),
                title: el.querySelector('.pdp-link a.link')?.textContent?.trim(),
                url: el.querySelector('.pdp-link a.link')?.getAttribute('href'),
                image_url: el.querySelector('.tile-image')?.getAttribute('src')
            }));
        })

        const results: MusemuStoreUpdate[] = [];

        for (const product of productsFromPageContext) {
            if (!product.id || !product.title || !product.url || !product.image_url) {
                throw new Error(`Missing required product attributes: ${JSON.stringify(product)}`);
            }

            const productFullUrl = new URL(product.url, 'https://store.muse.mu').href;

            results.push({
                type: UpdateType.MUSEMU_STORE,
                uniqueId: product.id,
                id: product.id,
                title: product.title,
                url: productFullUrl,
                image_url: product.image_url,
            });
        }

        return results;
    }
} 