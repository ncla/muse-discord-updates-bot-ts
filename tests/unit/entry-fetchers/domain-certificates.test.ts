import {expect, test} from 'vitest'
import {DomainCertificates} from "@/src/entry-fetchers/domain-certificates";
import {http, HttpResponse} from 'msw'
import {setupServer} from 'msw/node'
import {promises as fs} from "fs";
import path from "node:path";

test('it throws error when no domain is specified', async () => {
    // @ts-expect-error: TS2554
    const fetcherWithoutConstructParam = new DomainCertificates()
    await expect(fetcherWithoutConstructParam.fetch()).rejects.toThrow('Domain is not set')

    const fetcherWithUndefinedParam = new DomainCertificates(undefined)
    await expect(fetcherWithUndefinedParam.fetch()).rejects.toThrow('Domain is not set')
})

test('throws error on not OK response', async () => {
    const requestHandlers = [
        http.get('https://crt.sh/json', () => {
            return HttpResponse.json(
                {},
                {
                    status: 503,
                }
            )
        }),
    ]

    const server = setupServer(...requestHandlers)
    server.listen({ onUnhandledRequest: 'error' })

    const fetcher = new DomainCertificates('muse.mu')
    await expect(fetcher.fetch()).rejects.toThrow('Response status')

    server.close()
    server.resetHandlers()
})

test('it returns unique domain certificate entries on good response', async () => {
    const goodJsonResponse = await fs.readFile(
        path.join(__dirname, '../../__fixtures__/entry-fetchers/domain-certificates/good-json-response.json'),
        { encoding: 'utf-8' }
    );

    const requestHandlers = [
        http.get('https://crt.sh/json', () => {
            return HttpResponse.json(
                JSON.parse(goodJsonResponse),
                {
                    status: 200,
                }
            )
        }),
    ]

    const server = setupServer(...requestHandlers)
    server.listen({ onUnhandledRequest: 'error' })

    const fetcher = new DomainCertificates('muse.mu')
    const result = await fetcher.fetch()

    expect(result).toMatchSnapshot()

    server.close()
    server.resetHandlers()
})