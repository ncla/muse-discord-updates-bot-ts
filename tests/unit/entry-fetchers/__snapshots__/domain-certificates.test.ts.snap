// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`it returns unique domain certificate entries on good response 1`] = `
[
  {
    "author": {
      "id": null,
      "image_url": null,
      "name": null,
    },
    "content": null,
    "created_at": null,
    "id": "muse.mu",
    "image_url": null,
    "parent_id": null,
    "parent_title": null,
    "title": null,
    "type": "DOMAIN_CERTIFICATE",
    "uniqueId": "muse.mu",
    "url": null,
  },
  {
    "author": {
      "id": null,
      "image_url": null,
      "name": null,
    },
    "content": null,
    "created_at": null,
    "id": "forum.muse.mu",
    "image_url": null,
    "parent_id": null,
    "parent_title": null,
    "title": null,
    "type": "DOMAIN_CERTIFICATE",
    "uniqueId": "forum.muse.mu",
    "url": null,
  },
  {
    "author": {
      "id": null,
      "image_url": null,
      "name": null,
    },
    "content": null,
    "created_at": null,
    "id": "www.muse.mu",
    "image_url": null,
    "parent_id": null,
    "parent_title": null,
    "title": null,
    "type": "DOMAIN_CERTIFICATE",
    "uniqueId": "www.muse.mu",
    "url": null,
  },
  {
    "author": {
      "id": null,
      "image_url": null,
      "name": null,
    },
    "content": null,
    "created_at": null,
    "id": "ot-willofthepeople.muse.mu",
    "image_url": null,
    "parent_id": null,
    "parent_title": null,
    "title": null,
    "type": "DOMAIN_CERTIFICATE",
    "uniqueId": "ot-willofthepeople.muse.mu",
    "url": null,
  },
  {
    "author": {
      "id": null,
      "image_url": null,
      "name": null,
    },
    "content": null,
    "created_at": null,
    "id": "willofthepeople.muse.mu",
    "image_url": null,
    "parent_id": null,
    "parent_title": null,
    "title": null,
    "type": "DOMAIN_CERTIFICATE",
    "uniqueId": "willofthepeople.muse.mu",
    "url": null,
  },
  {
    "author": {
      "id": null,
      "image_url": null,
      "name": null,
    },
    "content": null,
    "created_at": null,
    "id": "usstore.muse.mu",
    "image_url": null,
    "parent_id": null,
    "parent_title": null,
    "title": null,
    "type": "DOMAIN_CERTIFICATE",
    "uniqueId": "usstore.muse.mu",
    "url": null,
  },
  {
    "author": {
      "id": null,
      "image_url": null,
      "name": null,
    },
    "content": null,
    "created_at": null,
    "id": "store.muse.mu",
    "image_url": null,
    "parent_id": null,
    "parent_title": null,
    "title": null,
    "type": "DOMAIN_CERTIFICATE",
    "uniqueId": "store.muse.mu",
    "url": null,
  },
  {
    "author": {
      "id": null,
      "image_url": null,
      "name": null,
    },
    "content": null,
    "created_at": null,
    "id": "tumblr.muse.mu",
    "image_url": null,
    "parent_id": null,
    "parent_title": null,
    "title": null,
    "type": "DOMAIN_CERTIFICATE",
    "uniqueId": "tumblr.muse.mu",
    "url": null,
  },
  {
    "author": {
      "id": null,
      "image_url": null,
      "name": null,
    },
    "content": null,
    "created_at": null,
    "id": "simulationtheory.muse.mu",
    "image_url": null,
    "parent_id": null,
    "parent_title": null,
    "title": null,
    "type": "DOMAIN_CERTIFICATE",
    "uniqueId": "simulationtheory.muse.mu",
    "url": null,
  },
  {
    "author": {
      "id": null,
      "image_url": null,
      "name": null,
    },
    "content": null,
    "created_at": null,
    "id": "pressure.muse.mu",
    "image_url": null,
    "parent_id": null,
    "parent_title": null,
    "title": null,
    "type": "DOMAIN_CERTIFICATE",
    "uniqueId": "pressure.muse.mu",
    "url": null,
  },
  {
    "author": {
      "id": null,
      "image_url": null,
      "name": null,
    },
    "content": null,
    "created_at": null,
    "id": "playlists.muse.mu",
    "image_url": null,
    "parent_id": null,
    "parent_title": null,
    "title": null,
    "type": "DOMAIN_CERTIFICATE",
    "uniqueId": "playlists.muse.mu",
    "url": null,
  },
  {
    "author": {
      "id": null,
      "image_url": null,
      "name": null,
    },
    "content": null,
    "created_at": null,
    "id": "colouringbook.muse.mu",
    "image_url": null,
    "parent_id": null,
    "parent_title": null,
    "title": null,
    "type": "DOMAIN_CERTIFICATE",
    "uniqueId": "colouringbook.muse.mu",
    "url": null,
  },
  {
    "author": {
      "id": null,
      "image_url": null,
      "name": null,
    },
    "content": null,
    "created_at": null,
    "id": "stageverse.muse.mu",
    "image_url": null,
    "parent_id": null,
    "parent_title": null,
    "title": null,
    "type": "DOMAIN_CERTIFICATE",
    "uniqueId": "stageverse.muse.mu",
    "url": null,
  },
  {
    "author": {
      "id": null,
      "image_url": null,
      "name": null,
    },
    "content": null,
    "created_at": null,
    "id": "somethinghuman.muse.mu",
    "image_url": null,
    "parent_id": null,
    "parent_title": null,
    "title": null,
    "type": "DOMAIN_CERTIFICATE",
    "uniqueId": "somethinghuman.muse.mu",
    "url": null,
  },
  {
    "author": {
      "id": null,
      "image_url": null,
      "name": null,
    },
    "content": null,
    "created_at": null,
    "id": "wontstanddown.muse.mu",
    "image_url": null,
    "parent_id": null,
    "parent_title": null,
    "title": null,
    "type": "DOMAIN_CERTIFICATE",
    "uniqueId": "wontstanddown.muse.mu",
    "url": null,
  },
  {
    "author": {
      "id": null,
      "image_url": null,
      "name": null,
    },
    "content": null,
    "created_at": null,
    "id": "*.muse.mu",
    "image_url": null,
    "parent_id": null,
    "parent_title": null,
    "title": null,
    "type": "DOMAIN_CERTIFICATE",
    "uniqueId": "*.muse.mu",
    "url": null,
  },
  {
    "author": {
      "id": null,
      "image_url": null,
      "name": null,
    },
    "content": null,
    "created_at": null,
    "id": "www.store.muse.mu",
    "image_url": null,
    "parent_id": null,
    "parent_title": null,
    "title": null,
    "type": "DOMAIN_CERTIFICATE",
    "uniqueId": "www.store.muse.mu",
    "url": null,
  },
  {
    "author": {
      "id": null,
      "image_url": null,
      "name": null,
    },
    "content": null,
    "created_at": null,
    "id": "origin.muse.mu",
    "image_url": null,
    "parent_id": null,
    "parent_title": null,
    "title": null,
    "type": "DOMAIN_CERTIFICATE",
    "uniqueId": "origin.muse.mu",
    "url": null,
  },
  {
    "author": {
      "id": null,
      "image_url": null,
      "name": null,
    },
    "content": null,
    "created_at": null,
    "id": "ai.muse.mu",
    "image_url": null,
    "parent_id": null,
    "parent_title": null,
    "title": null,
    "type": "DOMAIN_CERTIFICATE",
    "uniqueId": "ai.muse.mu",
    "url": null,
  },
  {
    "author": {
      "id": null,
      "image_url": null,
      "name": null,
    },
    "content": null,
    "created_at": null,
    "id": "film.muse.mu",
    "image_url": null,
    "parent_id": null,
    "parent_title": null,
    "title": null,
    "type": "DOMAIN_CERTIFICATE",
    "uniqueId": "film.muse.mu",
    "url": null,
  },
  {
    "author": {
      "id": null,
      "image_url": null,
      "name": null,
    },
    "content": null,
    "created_at": null,
    "id": "earlyaccess.muse.mu",
    "image_url": null,
    "parent_id": null,
    "parent_title": null,
    "title": null,
    "type": "DOMAIN_CERTIFICATE",
    "uniqueId": "earlyaccess.muse.mu",
    "url": null,
  },
  {
    "author": {
      "id": null,
      "image_url": null,
      "name": null,
    },
    "content": null,
    "created_at": null,
    "id": "board.muse.mu",
    "image_url": null,
    "parent_id": null,
    "parent_title": null,
    "title": null,
    "type": "DOMAIN_CERTIFICATE",
    "uniqueId": "board.muse.mu",
    "url": null,
  },
  {
    "author": {
      "id": null,
      "image_url": null,
      "name": null,
    },
    "content": null,
    "created_at": null,
    "id": "videoquiz.muse.mu",
    "image_url": null,
    "parent_id": null,
    "parent_title": null,
    "title": null,
    "type": "DOMAIN_CERTIFICATE",
    "uniqueId": "videoquiz.muse.mu",
    "url": null,
  },
  {
    "author": {
      "id": null,
      "image_url": null,
      "name": null,
    },
    "content": null,
    "created_at": null,
    "id": "www.usstore.muse.mu",
    "image_url": null,
    "parent_id": null,
    "parent_title": null,
    "title": null,
    "type": "DOMAIN_CERTIFICATE",
    "uniqueId": "www.usstore.muse.mu",
    "url": null,
  },
]
`;
